docker/Dockerfile

```diff
@@ -0,0 +1,71 @@
+# SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+# SPDX-License-Identifier: AGPL-3.0-only
+#
+# This Dockerfile uses features which are only available in BuildKit - see
+# https://docs.docker.com/go/buildkit/ for more information.
+#
+# To build the image, run `docker build` command from the root of the
+# repository:
+#
+#    DOCKER_BUILDKIT=1 docker build -f docker/Dockerfile .
+
+## Stage 0: Base image with only yarn and package.json
+FROM docker.io/node:16-alpine as base
+# Add tini to handle signals
+# https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#handling-kernel-signals
+USER  root
+RUN apk add --no-cache tini
+ENTRYPOINT ["tini"]
+
+USER node
+WORKDIR /usr/src/app
+
+COPY --chown=node .yarn .yarn
+COPY --chown=node package.json yarn.lock .yarnrc.yml ./
+
+
+## Stage 1: Code with all dependencies
+FROM base as code-with-deps
+USER node
+WORKDIR /usr/src/app
+
+# Install dependencies first to not invalidate the cache on every source change
+RUN --mount=type=cache,sharing=locked,uid=1000,gid=1000,target=/tmp/.yarn \
+    YARN_CACHE_FOLDER=/tmp/.yarn yarn install --immutable
+
+COPY --chown=node nest-cli.json tsconfig.json tsconfig.build.json ./
+COPY --chown=node src src
+
+
+## Stage 2a: Dev config files and tests
+FROM code-with-deps as development
+USER node
+WORKDIR /usr/src/app
+
+COPY --chown=node .eslintrc.js eslint-local-rules.js .prettierrc jest-e2e.json  ./
+COPY --chown=node test test
+
+CMD ["node", "-r", "ts-node/register", "src/main.ts"]
+
+## Stage 2b: Compile TypeScript
+FROM code-with-deps as builder
+USER node
+WORKDIR /usr/src/app
+
+RUN yarn run build
+
+
+## Stage 3: Final image, only production dependencies
+FROM base as prod
+
+USER node
+WORKDIR /usr/src/app
+ENV NODE_ENV=production
+
+COPY --chown=node --from=builder /usr/src/app/dist ./
+
+RUN --mount=type=cache,sharing=locked,uid=1000,gid=1000,target=/tmp/.yarn \
+    YARN_CACHE_FOLDER=/tmp/.yarn yarn install --immutable && \
```

Comment:
This install is superfluous as the `yarn workspaces focus` already does an install. See [the docs](https://yarnpkg.com/cli/workspaces/focus). So this install just adds all development dependencies as well (which is not what we want in the prod image).